apps/sms/services/js/activity/service.js

```diff
@@ -0,0 +1,93 @@
+/* global bridge */
+
+(function(exports) {
+  'use strict';
+
+  /**
+   * Name of the service that is responsible for managing activities.
+   * @type {string}
+   */
+  const SERVICE_NAME = 'activity-service';
+
+  /**
+   * Name of the system message that is leveraged by activities.
+   * @type {string}
+   */
+  const SYSTEM_MESSAGE_NAME = 'activity';
+
+  /**
+   * Reference to active bridge service instance.
+   * @type {Service}
+   */
+  var service = null;
+
+  /**
+   * Reference to currently active activity request.
+   * @type {ActivityRequestHandler}
+   */
+  var activityRequest = null;
+
+  /**
+   * Posts activity request result.
+   * @param {*} result Data to post as activity request result.
+   */
+  function onPostResult(result) {
+    if (!activityRequest) {
+      throw new Error('There is no any activity request to post result to!');
+    }
+
+    activityRequest.postResult(result || { success: true });
+
+    activityRequest = null;
+  }
+
+  /**
+   * Posts activity request error.
+   * @param {*} error Error to post as activity request error.
+   */
+  function onPostError(error) {
+    if (!activityRequest) {
+      throw new Error('There is no any activity request to post error to!');
+    }
+
+    activityRequest.postError(error);
+
+    activityRequest = null;
+  }
+
+  /**
+   * Handler that fires once app receives activity request via system message.
+   * @param {ActivityRequestHandler} request Activity request.
+   */
+  function onActivityRequest(request) {
+    activityRequest = request;
+
+    // Talk to Wilson to understand how to fix this.
+    setTimeout(() => {
+      service.broadcast('activity-request', request.source);
+    });
+  }
+
+  var ActivityService = {
+    /**
+     * Initialized activity service bridge.
+     */
+    init() {
+      service = bridge.service(SERVICE_NAME);
+
+      service.method('postResult', onPostResult);
+      service.method('postError', onPostError);
+
+      navigator.mozSetMessageHandler(SYSTEM_MESSAGE_NAME, onActivityRequest);
+    },
+
+    /**
+     * Checks if activity service has pending activity request.
+     */
+    hasPendingRequest() {
+      return navigator.mozHasPendingMessage(SYSTEM_MESSAGE_NAME);
```

Comment:
Just want to hide everything related to activity&system message dependency inside service wrapper.